File: core-cpp/src/runtime/dispatch.cpp
// Type-id → factory registry for typed message dispatch. See
// dispatch.hpp for the public API (register_message_type /
// lookup_message_class / unregister_message_type) and error model.

#include "oigtl/runtime/dispatch.hpp"

#include <algorithm>
#include <string>

namespace oigtl::runtime {

void Registry::register_message_type(std::string type_id,
                                     RoundTripFn fn,
                                     bool replace) {
    const auto it = table_.find(type_id);
    if (it != table_.end()) {
        if (it->second == fn) {
            // Idempotent: registering the same (type_id, fn) pair
            // twice is a no-op, matching Python / TypeScript.
            return;
        }
        if (!replace) {
            throw RegistryConflictError(
                "type_id \"" + type_id +
                "\" is already registered; pass replace=true to "
                "replace it, or pick a different type_id");
        }
    }
    table_.insert_or_assign(std::move(type_id), fn);
}

RoundTripFn Registry::unregister_message_type(std::string_view type_id) {
    // One allocating copy; see the header note about heterogeneous
    // lookup being a C++20 feature.
    const auto it = table_.find(std::string(type_id));
    if (it == table_.end()) {
        return nullptr;
    }
    const RoundTripFn prior = it->second;
    table_.erase(it);
    return prior;
}

RoundTripFn Registry::lookup_message_class(
    std::string_view type_id) const noexcept {
    const auto it = table_.find(std::string(type_id));
    return (it == table_.end()) ? nullptr : it->second;
}

std::vector<std::string> Registry::registered_types() const {
    std::vector<std::string> out;
    out.reserve(table_.size());
    for (const auto& [key, _] : table_) {
        out.push_back(key);
    }
    std::sort(out.begin(), out.end());
    return out;
}

}  // namespace oigtl::runtime

File: core-ts/src/runtime/dispatch.ts
/**
 * Public registry of message body classes, keyed by wire `type_id`.
 *
 * Built-in message classes (generated in `messages/`) and third-party
 * extension classes (PLUS `TRACKEDFRAME`, vendor-specific types,
 * etc.) share this single table — there is no private registration
 * back door. Registering an extension makes it decode through the
 * exact same codec path the built-ins use.
 *
 * Mirrors the Python `oigtl.messages.registry` API.
 */

import type { Header } from "./header.js";

/**
 * Generated message classes conform to this shape: a constructor
 * with a static `TYPE_ID` and a static `unpack(bytes)` factory.
 * They may also expose `BODY_SIZE` for fixed-body types and a
 * `pack(): Uint8Array` instance method (required for round-trip
 * encoding).
 */
export interface MessageCtor<T = unknown> {
  readonly TYPE_ID: string;
  unpack(bytes: Uint8Array): T;
}

/**
 * Raised when `registerMessageType` would silently overwrite an
 * existing entry. Pass `override: true` to force replacement.
 */
export class RegistryConflictError extends Error {
  constructor(typeId: string, existingName: string) {
    super(
      `type_id ${JSON.stringify(typeId)} is already registered to ` +
        `${existingName}; pass { override: true } to replace it, or ` +
        `pick a different type_id`,
    );
    this.name = "RegistryConflictError";
  }
}

const _registry = new Map<string, MessageCtor>();

/**
 * Register `ctor` as the body class for its declared `TYPE_ID`.
 *
 * After registration, `unpackMessage` / `unpackEnvelope` dispatch
 * wire messages whose header `typeId` matches through
 * `ctor.unpack(body)`.
 *
 * @param ctor - The body class. Must expose `TYPE_ID` (up to 12
 *   ASCII chars per the OpenIGTLink spec) and `unpack(bytes)`.
 * @param opts.override - If true, silently replace any existing
 *   entry. If false (default) and `ctor.TYPE_ID` is already
 *   registered to a different class, throws
 *   {@link RegistryConflictError}. Re-registering the same
 *   `(TYPE_ID, ctor)` pair is always idempotent.
 *
 * @throws {RegistryConflictError} `ctor.TYPE_ID` is already bound
 *   to a different class and `override` is false.
 */
export function registerMessageType(
  ctor: MessageCtor,
  opts: { override?: boolean } = {},
): void {
  const existing = _registry.get(ctor.TYPE_ID);
  if (existing !== undefined && existing !== ctor && !opts.override) {
    const existingName =
      (existing as unknown as { name?: string }).name ?? "anonymous";
    throw new RegistryConflictError(ctor.TYPE_ID, existingName);
  }
  _registry.set(ctor.TYPE_ID, ctor);
}

/**
 * Remove and return the class registered for `typeId`, if any.
 *
 * Primarily useful in tests that need to undo a registration
 * between cases. Production code should not need this.
 */
export function unregisterMessageType(typeId: string): MessageCtor | undefined {
  const prior = _registry.get(typeId);
  _registry.delete(typeId);
  return prior;
}

/**
 * Return the class registered for `typeId`, or `undefined`.
 *
 * Unlike `registerMessageType`, this never throws — callers use
 * `undefined` as the signal to fall back to a raw-body path
 * (see {@link RawBody} in `net/envelope.ts`).
 */
export function lookupMessageClass(typeId: string): MessageCtor | undefined {
  return _registry.get(typeId);
}

/** Sorted list of every currently-registered type_id. */
export function registeredTypes(): string[] {
  return [...(_registry.keys())].sort();
}

/** Number of currently-registered types (tooling / diagnostics). */
export function registrySize(): number {
  return _registry.size;
}

/** Context passed to every typed constructor for error attribution. */
export interface DispatchContext {
  header: Header;
}
